### ISSUE TITLE
FTBFS with JDK9

### ISSUE THREAD

**Issue by hansjoachim** [junit-team/junit#1159]
Hello

Ubuntu recently added an early snapshot of JDK9 to their current development release, so I'm testing it out on various projects. I tried to build JUnit (latest HEAD 989be00be3baecb3e47bc95ba981d642bb9eca81), but it failed to build from source.
```
$ mvn clean compile -e
Picked up JAVA_TOOL_OPTIONS: -javaagent:/usr/share/java/jayatanaag.jar 
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building JUnit 4.13-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ junit ---
[INFO] Deleting /home/ubuntu/junit/target
[INFO] 
[INFO] --- maven-enforcer-plugin:1.3.1:enforce (enforce-versions) @ junit ---
[INFO] 
[INFO] --- replacer:1.5.3:replace (default) @ junit ---
[INFO] Replacement run on 1 file.
[INFO] 
[INFO] --- maven-resources-plugin:2.3:resources (default-resources) @ junit ---
[INFO] Using 'ISO-8859-1' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Copying 1 resource
[INFO] 
[INFO] --- maven-compiler-plugin:3.3:compile (default-compile) @ junit ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 195 source files to /home/ubuntu/junit/target/classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 54.709s
[INFO] Finished at: Sun Jun 07 20:23:04 CEST 2015
[INFO] Final Memory: 9M/22M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project junit: Compilation failure -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project junit: Compilation failure
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:213)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:502)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:915)
	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:129)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 19 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

The build was attempted with Maven 3.0.5-5 / OpenJDK version 9~b64-1ubuntu1 available on Ubuntu Wily. I'm not yet 100% sure whether this is a bug in JUnit or whether it could be in the JDK (or somewhere else). Has anyone else looked into or experimented with JDK9?

With some minor tweaks, I was able to compile the maven-compiler-plugin locally though, so I suppose that should work.

**Comment by kcooney** [comment]
FYI: We have no plans to do a release of JUnit that requires JDK 9. JUnit 5 will require JDK 6 or 7.

**Comment by hansjoachim** [comment]
@kcooney: Yes, I'm aware of this. I understand that due to the number of dependents using JUnit such a high version requirement wouldn't work, so I won't argue that point.:)

I believe it should be possible to build the project using a more recent JDK though.

**Comment by Tibor17** [comment]
@hansjoachim 
I guess the sources of junit are okay, because the compiler did not list any source failure. This looks like maven bug. Please send email to Maven Developer mailing list. We deal with JDK 9.

**Comment by rschmitt** [comment]
Yeah, this is almost certainly a Maven issue. Ant had a similar issue when Java 8 come out. These build systems almost always have a hard-coded switch statement lurking somewhere that spells out individual major versions of Java.

**Comment by Stephan202** [comment]
Nope, this is due to [JEP 182](http://openjdk.java.net/jeps/182). Proof:

    $ mvn -X clean compile >/dev/null ; cat target/classes/org.codehaus.plexus.compiler.javac.JavacCompiler* | xargs javac
    warning: [options] bootstrap class path not set in conjunction with -source 1.5
    error: Source option 1.5 is no longer supported. Use 1.6 or later.
    error: Target option 1.5 is no longer supported. Use 1.6 or later.

**Comment by rschmitt** [comment]
Great catch. @hansjoachim: try it on the junit5 branch, I guess?

**Comment by hansjoachim** [comment]
@Stephan202 Ah, thanks. :) I've seen similar error messages when attempting to build other projects, but I got confused when I only got a weird stacktrace. Also wasn't aware there was a JEP on it. Good to know.

Just tried the junit5 branch, and was able to run `mvn clean install` successfully. So all in all, I consider that fixed. Does someone have the power to link this issue to the junit5 branch and/or milestone?
